{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding CRISPR arrays\n",
    "## Specification\n",
    "## Program Name: randomizedMotifSearch.py\n",
    "\n",
    "## Input/Output: STDIN, STDOUT\n",
    "\n",
    "## Options:\n",
    "\n",
    "- -i iterations (int)\n",
    "- -k motif length (int)\n",
    "- -p pseudocount (float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ex: python randomizedMotifSearch.py -i 1000 -k 13 -p 1 < input.fa > output.out\n",
    "\n",
    "Inspection Notebook example:\n",
    "main ( \"input.fa\", options = [ \"-i 1000\", \"-k 13\", \"-p 1\"] )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Background\n",
    "CRISPR arrays exist in most archaeal species and many bacterial species. The arrays are encoded in chromosomal DNA as direct repeats (DR) that flank sequences (spacers) that provide immunity against invading plasmids, viral sequences, and other foreign DNA .. and RNA. These arrays can be short, comprising 1-2 spacers, or they can contain hundreds of spacer elements. The arrays are punctuated by these DR sequences. The promoter element seems to be upstream of the array, and initiates transcription of the associated RNA strand that is then cleaved and those spacers are loaded into protein machines that provide specific targeting of the associated invader should it be found. That targeting causes cleavage of the foreign DNA in known cases.\n",
    "\n",
    "There are three main classes of CRISPR systems; these are type I,II and III, and are established based on the associated protein machines - actually the specific genes that encode for them. In the archaea, only type I and III have been found so far, while in bacteria, all three types have been seen. One of those types, type II, exists in a few species, including _Streptococcus pyogenes_, which encodes for a type II system. The specific gene that is used to classify this system is named Cas9.\n",
    "\n",
    "We see cases where Cas genes, and sometimes CRISPR arrays, exist in a viral sequence. This, of course, opens many questions about the role of the CRISPR system with respect to viruses - why would a virus carry an immune defense that targets viral sequence? The existence of these arrays suggests that the CRISPR system itself might be mobile. How does the system get established, and how does the surrounding transcription machinery get encoded on the chromosome?\n",
    "\n",
    "For this assignment, we will work to find the promoter motif. We would expect that this promoter motif and an associated B recognition element (BRE) would be present in archaeal species. What is that motif, and where is it located relative to transcription start?\n",
    "\n",
    "We know little about the sequence specifics of the promoter element, though it seems to be about 13 nucleotides in length. In Pyrobaculum species, genes are packed quite tightly, so we can look upstream of the array by about 50 bases to see if a common motif sequence might exist. We need to assume that some mutations can be tolerated in that sequence.\n",
    "\n",
    "## Null model\n",
    "In prior years, we included an option that scrambled the input sequences before running to find a best score. This would effectively eliminate any signal that might be present in the data, while preserving the overall composition of the sequences. We would then compare the best score achieved without scrambling to the score achieved with scrambling. The information that remains in a scrambled set of sequences is only the base level composition of the sequences:\n",
    "\n",
    "$Q=Pr_{s\\in\\lbrace{ACTG}\\rbrace}(s)=\\frac{count(s)}{N}$\n",
    "\n",
    "We can then use relative entropy to find a score relative to the base level composition by:\n",
    "\n",
    "$REscore = \\sum_{i=0}^{cols}\\sum^{j \\in ACGT}P_{i}(j)\\log_{2}(\\frac{P_{i}(j)}{Q(j)} )$\n",
    "\n",
    "## Assignment\n",
    "__Write a BME205 style program (.py file)__ that is based on the Randomized Motif Search that presented in class/videos and is described in Chapter 2 of the text. Your program should accept a fasta file that contains multiple fasta records as input (STDIN), and your program will then output (STDOUT) the consensus sequence and associated profile score. The score will be the sum of encoding costs (entropies) across each position in the final profile. We will use pseudocounts for this assignment, as described in the text ( p. 86-89), and we will need an option to provide the number of pseudocounts (-p).\n",
    "\n",
    "__Produce a notebook that should be fully runable for use during inspections.__\n",
    "\n",
    "I find it easier to accumulate counts rather than calculating a probability based profile. This will also make it a bit easier to deal with pseudocounts when scoring your count-based profile.\n",
    "\n",
    "$score = \\sum_{i=0}^{cols}\\sum^{j \\in ACGT}P_{i}(j)\\log_{2}(\\frac{P_{i}(j)}{Q(j)} )$\n",
    "\n",
    "Randomized Motif Search (p. 93) can be easily trapped in local minima, so we will need to iterate some number of times to find a trajectory that produces the best results. This will need to be an option (-i) that establishes the iteration number.\n",
    "\n",
    "We also don't know the appropriate motif length that we are looking for, so we need an option (-k) to specify this.\n",
    "\n",
    "The full command line string should then look something like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "randomizedMotifSearch.py -i=100000 -p=1 -k=13 <somefile.fa >someOutputFile.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "your program will need to use standard OO style, include docstrings and be well commented. It must run! \n",
    "\n",
    "I will provide a few fasta files that present the upstream 50 bases from species of the Pyrobaculum group. It is possible that the promotor+BRE that we are looking for is conserved across these species, maybe not perfectly though. Providing more data to a single run may be useful.\n",
    "\n",
    "I do have expression data for these species, so we know that not all of these arrays are functional - at least under the conditions when I grew them.\n",
    "\n",
    "Your program should output the final score achieved, along with the associated consensus motif.\n",
    "\n",
    "## Extra Credit\n",
    "The following optional features would be useful and considered for a combined extra 5 points:\n",
    "\n",
    "1) -g Use Gibbs sampling to find the optimal consensus motif.\n",
    "\n",
    "2) -m Print the specific motif and the name of the contributing sequence for each of the participating fasta records.\n",
    "\n",
    "Each of these options should only be true when the flag is specified by the user (ie. you shouldn't have to explicitly state True/False on the commandline).\n",
    "\n",
    "Submit your working program to Canvas.\n",
    "\n",
    "## Notes:\n",
    "\n",
    "1) It is likely that we will need to amend this assignment as we work with it.\n",
    "\n",
    "2) Sept-2021 eliminated the need for scrambling sequences by using relative entropy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspection intro\n",
    "\n",
    "- What data structure did you use to hold your sequence motifs (DNA)? \n",
    "- Is your profile  structure based on counts or frequencies? \n",
    "- Where did you implement iterations (mostly why did you choose this)?\n",
    "- If you implemented Gibbs Sampling, how did you implement recomputing of the profile and motif?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspection Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<>:25: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<>:25: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "<ipython-input-3-2bae87261092>:25: SyntaxWarning: \"is\" with a literal. Did you mean \"==\"?\n",
      "  if self.fname is '':\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGAACTGT   2.984359622300305\n"
     ]
    }
   ],
   "source": [
    "# Randomized Motif Search\n",
    "# input: A fasta file with multiple fasta records\n",
    "# returns the consensus sequence and associated profile score\n",
    "# Groupmates: Alan Faber, Sarah Ghasemi\n",
    "# Author: Jeffrey Jacob\n",
    "\n",
    "import sys\n",
    "from random import random\n",
    "import random\n",
    "import math\n",
    "class FastAreader:\n",
    "    ''' This class reads in the FastA file and provides the main the header and sequence in the FastA file'''\n",
    "    def __init__(self, fname=''):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    "    def doOpen(self):\n",
    "        ''' Opens the file in stdin '''\n",
    "        if self.fname is '':\n",
    "            return sys.stdin\n",
    "        else:\n",
    "            return open(self.fname)\n",
    "    def readFasta(self):\n",
    "        ''' Reads the file and returns the header and sequence one by one '''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        with self.doOpen() as fileH:\n",
    "            header = ''\n",
    "            sequence = ''\n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>'):\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    "            for line in fileH:\n",
    "                if line.startswith('>'):\n",
    "                    yield header, sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else:\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        yield header, sequence\n",
    "\n",
    "class CommandLine():\n",
    "    ''' Deals with the commandLine options provided and sets the values to be used by main and other classes '''\n",
    "    def __init__(self, inOpts=None):\n",
    "        ''' Contains all the information that either saves the information given by the user, or uses the default values'''\n",
    "        import argparse\n",
    "        self.parser = argparse.ArgumentParser(description='This program find the motif with the highest relative entropy score', add_help=True)\n",
    "        self.parser.add_argument('-i', '--iterations', type=int, action='store', default=1000, dest='iterations',\n",
    "                                 help='Provide the number of runs the motif search should take, the default is 1000 iterations.') #handles the iterations option provided by the user, or set as the default\n",
    "        self.parser.add_argument('-p', '--pseudoCount', type=float, action='store', default=1,dest='pseudoCount',\n",
    "                                 help='Provide the psuedo count, the default is 1 pseudo count.') #handles the pseudo count option provided by the user, or set as the default\n",
    "        self.parser.add_argument('-k', '--kmerLength', type=int, action='store', default=13, dest='kmerLength',\n",
    "                                 help='Provide the kmer length') #handles the kmer length option provided by the user, or set as the default\n",
    "        if inOpts is None:\n",
    "            self.args = self.parser.parse_args()\n",
    "        else :\n",
    "            self.args = self.parser.parse_args(inOpts)\n",
    "\n",
    "class Calculations:\n",
    "    ''' The calculations class deals with all the processing related to the sequences '''\n",
    "    def __init__(self):\n",
    "        ''' Description: Initializes all the values that will be utilized within the class '''\n",
    "        ''' Input: The method does not have any inputs '''\n",
    "        ''' Output: The method initializes certain values that will be changed in future methods '''\n",
    "        self.aCount = 0; self.tCount = 0; self.cCount = 0; self.gCount = 0 # Initializes the counts of all the nucleotides\n",
    "        self.aNull = 0; self.tNull = 0; self.cNull = 0; self.gNull = 0 # Initializes the null values of all the nucleotides\n",
    "        self.total = 0 \n",
    "        self.profArray = [] \n",
    "\n",
    "    def createNull(self, a, t, c, g, pCount):\n",
    "        ''' Description: Calculates the Null Model '''\n",
    "        ''' Input: This method uses the number of A, T, C and G and the pseudo count '''\n",
    "        ''' Output: The method saves the null values of each of the nucleotides to be used in other methods '''\n",
    "        self.aCount += a; self.tCount += t; self.cCount += c; self.gCount += g \n",
    "        self.total = a + t + c + g + (pCount*4) \n",
    "        self.aNull = ((a+pCount)/self.total); self.tNull = ((t+pCount)/self.total); self.cNull = ((c+pCount)/self.total); self.gNull = ((g+pCount)/self.total) # Calculates the null probabilities of each of the nucleotides\n",
    "\n",
    "    def randomMotif(self, seqList, kmerLength):\n",
    "        ''' Description: Retrieves random motifs of a certain length in each of the sequences '''\n",
    "        ''' Input: This method uses a list of sequences and the length of the k-mer which is specified by the user '''\n",
    "        ''' Output: The method returns a random motif within each of the sequences '''\n",
    "        totalLength = len(seqList[0])\n",
    "        ranList = [random.randint(0, totalLength-(kmerLength+1)) for i in seqList] \n",
    "        motifList = []; count = 0\n",
    "        for i in seqList:\n",
    "            motifList.append(i[ranList[count]:ranList[count]+kmerLength]) # Finds and saves the selected motif into a list of motifs\n",
    "            count+=1\n",
    "        return motifList\n",
    "\n",
    "    def createProfile(self, motifList, kmerLength, pseudoCount):\n",
    "        ''' Description: This method determines the profile given the list of motifs and length of the k-mer '''\n",
    "        ''' Input: The input for this method is the list of motif, the length of the k-mer, and the pseudo count '''\n",
    "        ''' Output: The method returns the profile '''\n",
    "        self.profArray = []; subList = []\n",
    "        for i in range(kmerLength): # Creates an empty list of lists\n",
    "            subList.append([])\n",
    "        numMotif = len(motifList)\n",
    "        aProfList = []; tProfList = []; cProfList = []; gProfList = [];\n",
    "        for i in range(kmerLength): # Creates a 2D array to create the profile by counting the number of nucleotides in specific positions\n",
    "            for j in range(numMotif):\n",
    "                subList[i].append(motifList[j][i]) \n",
    "            aProfList.append((subList[i].count(\"A\")+pseudoCount)/(numMotif+(pseudoCount*4)))\n",
    "            tProfList.append((subList[i].count(\"T\")+pseudoCount)/(numMotif+(pseudoCount*4)))\n",
    "            cProfList.append((subList[i].count(\"C\")+pseudoCount)/(numMotif+(pseudoCount*4)))\n",
    "            gProfList.append((subList[i].count(\"G\")+pseudoCount)/(numMotif+(pseudoCount*4)))\n",
    "        self.profArray.append(aProfList); self.profArray.append(tProfList); self.profArray.append(cProfList); self.profArray.append(gProfList) # Places the probabilities into a profile array\n",
    "        return self.profArray\n",
    "\n",
    "    def iterMotifs(self, seqList, kmerLength):\n",
    "        ''' Description: Determines all the motifs in a given set of sequences, and determines a motif with the highest score given the profile '''\n",
    "        ''' Input: The method uses the all the sequences and the length of the k-mer '''\n",
    "        ''' Output: The output is the final motif that has the highest value '''\n",
    "        totalLength = len(seqList[0])\n",
    "        finalMotif = []\n",
    "        for i in range(len(seqList)): # Iterates through all the sequences to find the best possible motif for each sequence\n",
    "            motifList = []\n",
    "            for j in range(totalLength-kmerLength): # Finds all the possible motifs in a given sequence\n",
    "                motifList.append(seqList[i][j:j+kmerLength])\n",
    "            motifDict = {}\n",
    "            for k in range(len(motifList)): # Iterates through all the motifs and calculates based on the profile\n",
    "                count = 1\n",
    "                for l in range(kmerLength): # Iterates through the length of the kmer and calculates it's score based on profile\n",
    "                    if (motifList[k][l]) == \"A\": count *=(self.profArray[0][l])\n",
    "                    elif (motifList[k][l]) == \"T\": count *= (self.profArray[1][l])\n",
    "                    elif (motifList[k][l]) == \"C\": count *= (self.profArray[2][l])\n",
    "                    elif (motifList[k][l]) == \"G\": count *= (self.profArray[3][l])\n",
    "                motifDict[motifList[k]] = count # Places the motif and it's correpsonding score in a dictionary\n",
    "            finalMotif.append(max(motifDict, key=motifDict.get)) # Determines the motif with the highest value and adds it to a list of motifs\n",
    "        return finalMotif\n",
    "\n",
    "    def relativeEntropy(self, profile, consensus):\n",
    "        ''' Description: Calculates the relative entropy of the consensus sequence '''\n",
    "        ''' Input: The function uses the profile of the final motif and the determined consensus '''\n",
    "        ''' Output: The method returns the relative entropy score of the consensus '''\n",
    "        score = 0\n",
    "        for i in range(len(consensus)): # Goes through all the nucleotides within the consensus and calculates the total relative entropy score\n",
    "            if consensus[i] == \"A\":\n",
    "                score += (profile[0][i]) * (math.log2((profile[0][i])/self.aNull))\n",
    "            elif consensus[i] == \"T\":\n",
    "                score += (profile[1][i]) * (math.log2((profile[1][i])/self.tNull))\n",
    "            elif consensus[i] == \"C\":\n",
    "                score += (profile[2][i]) * (math.log2((profile[2][i])/self.cNull))\n",
    "            elif consensus[i] == \"G\":\n",
    "                score += (profile[3][i]) * (math.log2((profile[3][i])/self.gNull))\n",
    "        return (score)\n",
    "\n",
    "def main (inFile, myCommandLine=None):\n",
    "    ''' Description: Runs the entire program by reading the command line and outputting the corresponding profile and relative entropy score '''\n",
    "    ''' Input: The input is the options given by the user through the commandLine '''\n",
    "    ''' Output: The main function prints out the consensus with the highest relative entropy score and its score into the specified output file '''\n",
    "    if myCommandLine is None:\n",
    "        myCommandLine = CommandLine()  # read options from the command line\n",
    "    else:\n",
    "        myCommandLine = CommandLine(\n",
    "            myCommandLine)  # interpret the list passed from the caller of main as the commandline.\n",
    "    \n",
    "    seqList = []\n",
    "    aCount = 0; tCount = 0; cCount = 0; gCount = 0\n",
    "    for head, seq in FastAreader(inFile).readFasta(): #iterates through all the sequences within the Fasta\n",
    "        processedSeq = Calculations()\n",
    "        seqList.append(seq)\n",
    "        aCount += seq.count(\"A\"); tCount += seq.count(\"T\"); cCount += seq.count(\"C\"); gCount += seq.count(\"G\")\n",
    "    processedSeq.createNull(aCount, tCount, cCount, gCount, myCommandLine.args.pseudoCount) # Determines the Null Model given the counts of nucleotides within the sequences\n",
    "    allEntropy = []; allConsensus = []\n",
    "    for i in range(myCommandLine.args.iterations): # Iterates through the number of user provided user iterations and outputs the relative entropy score along with its corresponding consensus\n",
    "        oldMotifList = processedSeq.randomMotif(seqList, myCommandLine.args.kmerLength) # Creates the initial motif list using the random motif generator\n",
    "        processedSeq.createProfile(oldMotifList, myCommandLine.args.kmerLength, myCommandLine.args.pseudoCount)\n",
    "        newMotifList = processedSeq.iterMotifs(seqList, myCommandLine.args.kmerLength)\n",
    "        while oldMotifList != newMotifList: # Determines the best motif list which is attained with the new and old motif list are the same\n",
    "            oldMotifList = newMotifList \n",
    "            profile = processedSeq.createProfile(oldMotifList, myCommandLine.args.kmerLength, myCommandLine.args.pseudoCount) # Finds the profile of the last motif list\n",
    "            newMotifList = processedSeq.iterMotifs(seqList, myCommandLine.args.kmerLength) # Saves the motif list to a newMotifList to compare\n",
    "        consensus = []\n",
    "        for i in range(myCommandLine.args.kmerLength): # Iterates through the length of the kmer and determines the consensus sequence\n",
    "            aVal = profile[0][i]; tVal = profile[1][i]; cVal = profile[2][i]; gVal = profile[3][i] \n",
    "            maxNuc = ([aVal, tVal, cVal, gVal]).index(max([aVal, tVal, cVal, gVal])) # Determines which nucleotide was in the highest frequence\n",
    "            if maxNuc == 0:nucVal = \"A\"\n",
    "            elif maxNuc == 1:nucVal = \"T\"\n",
    "            elif maxNuc == 2:nucVal = \"C\"\n",
    "            elif maxNuc == 3:nucVal = \"G\"\n",
    "            consensus.append(nucVal) # Appends the nucleotide to the consensus\n",
    "        entropyScore = processedSeq.relativeEntropy(profile, consensus) \n",
    "        allConsensus.append(consensus) # Appends the consenses to a list of consensus's \n",
    "        allEntropy.append(entropyScore) \n",
    "    finalEntropy = max(allEntropy) # Finds the highest entrpy score\n",
    "    finalConsensus = \"\".join(allConsensus[allEntropy.index(finalEntropy)]) \n",
    "    print(str(finalConsensus) + \"   \" + str(finalEntropy)) # Prints the consensus and it's entropy score to the output file\n",
    "\n",
    "if __name__== \"__main__\":\n",
    "    main(\"pogCrisprs\", [\"-i 1000\", \"-p 3\", \"-k 8\"])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "nteract": {
   "version": "0.28.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
